main-command/src/main/scala/sbt/FileTreeViewConfig.scala

```diff
@@ -0,0 +1,164 @@
+/*
+ * sbt
+ * Copyright 2011 - 2017, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under BSD-3-Clause license (see LICENSE)
+ */
+
+package sbt
+import sbt.Watched.WatchSource
+import sbt.internal.io.{ HybridPollingFileRepository, WatchServiceBackedObservable, WatchState }
+import sbt.io._
+import FileTreeDataView.{ Observable, Observer }
+import sbt.util.Logger
+
+import scala.concurrent.duration._
+
+/**
+ * Configuration for viewing and monitoring the file system.
+ */
+final class FileTreeViewConfig private (
+    val newDataView: () => FileTreeDataView[StampedFile],
+    val newMonitor: (
+        FileTreeDataView[StampedFile],
+        Seq[WatchSource],
+        Logger
+    ) => FileEventMonitor[StampedFile]
+)
+object FileTreeViewConfig {
+  private implicit class RepositoryOps(val repository: FileRepository[StampedFile]) {
+    def register(sources: Seq[WatchSource]): Unit = sources foreach { s =>
+      repository.register(s.base.toPath, if (s.recursive) Integer.MAX_VALUE else 0)
+    }
+  }
+
+  /**
+   * Create a new FileTreeViewConfig. This factory takes a generic parameter, T, that is bounded
+   * by [[FileTreeDataView]][StampedFile]. The reason for this is to ensure that a
+   * [[FileTreeDataView]] that is instantiated by [[FileTreeViewConfig.newDataView]] can be passed
+   * into [[FileTreeViewConfig.newMonitor]] without constraining the type of view to be
+   * [[FileTreeDataView]][StampedFile].
+   * @param newDataView create a new [[FileTreeDataView]]. This value may be cached in a global
+   *                    attribute
+   * @param newMonitor create a new [[FileEventMonitor]] using the [[FileTreeDataView]] created
+   *                   by newDataView
+   * @tparam T the subtype of [[FileTreeDataView]] that is returned by [[FileTreeViewConfig.newDataView]]
+   * @return a [[FileTreeViewConfig]] instance.
+   */
+  def apply[T <: FileTreeDataView[StampedFile]](
+      newDataView: () => T,
+      newMonitor: (T, Seq[WatchSource], Logger) => FileEventMonitor[StampedFile]
+  ): FileTreeViewConfig =
+    new FileTreeViewConfig(
+      newDataView,
+      (view: FileTreeDataView[StampedFile], sources: Seq[WatchSource], logger: Logger) =>
+        newMonitor(view.asInstanceOf[T], sources, logger)
+    )
+
+  /**
+   * Provides a [[FileTreeViewConfig]] with semantics as close as possible to sbt 1.2.0. This means
+   * that there is no file tree caching and the [[FileEventMonitor]] will use a [[WatchService]].
+   * @param antiEntropy the duration of the period after a path triggers a build for which it is
+   * @return a [[FileTreeViewConfig]] instance.
+   */
+  def `1.2.0`(pollingInterval: FiniteDuration, antiEntropy: FiniteDuration): FileTreeViewConfig =
```

Comment:
Not sure if this name is clear when it comes up as 

```scala
fileTreeViewConfig := FileTreeViewConfig.`1.2.0`(pollInterval.value, watchAntiEntropy.value)
```

Maybe `sbt1_2_compat`?